modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1028,4 +1087,357 @@ public Set<ReplicationGroupId> startedGroups() {
+                            + parameters.groupId() + "].";
+
+                    context.reservedForPrimary = true;
+                } else {
+                    context.reservedForPrimary = false;
+
+                    if (context.replicaState == ReplicaState.PRIMARY_ONLY) {
+                        stopReplica(parameters.groupId(), context, context.deferredStopOperation);
+                    }
+                }
+            }
+
+            return falseCompletedFuture();
+        }
+
+        private CompletableFuture<Boolean> onPrimaryExpired(PrimaryReplicaEventParameters parameters) {
+            if (localNodeId.equals(parameters.leaseholderId())) {
+                ReplicaLifecycleContext context = replicaContexts.get(parameters.groupId());
+
+                if (context != null) {
+                    synchronized (context) {
+                        context.reservedForPrimary = false;
+                    }
+                }
+            }
+
+            return falseCompletedFuture();
+        }
+
+        ReplicaLifecycleContext getContext(ReplicationGroupId groupId) {
+            return replicaContexts.computeIfAbsent(groupId,
+                    // Treat the absence in the map as STOPPED.
+                    k -> new ReplicaLifecycleContext(ReplicaState.STOPPED, nullCompletedFuture()));
+        }
+
+        /**
+         * Can possibly start replica if it's not running or is stopping.
+         *
+         * @param groupId Group id.
+         * @param startOperation Replica start operation.
+         * @param forcedAssignments Assignments to reset forcibly, if needed. Assignments reset is only available when replica is started.
+         */
+        CompletableFuture<Boolean> weakReplicaStart(
+                ReplicationGroupId groupId,
+                Supplier<CompletableFuture<Boolean>> startOperation,
+                @Nullable Assignments forcedAssignments
+        ) {
+            ReplicaLifecycleContext context = getContext(groupId);
+
+            synchronized (context) {
+                ReplicaState state = context.replicaState;
+
+                log.debug("Weak replica start [grp={}, state={}, future={}].", groupId, state, context.previousOperationFuture);
+
+                if (state == ReplicaState.STOPPED || state == ReplicaState.STOPPING) {
+                    return startReplica(groupId, context, startOperation);
+                } else if (state == ReplicaState.ASSIGNED) {
+                    if (forcedAssignments != null) {
+                        assert forcedAssignments.force() :
+                                format("Unexpected assignments to force [assignments={}, groupId={}].", forcedAssignments, groupId);
+
+                        replicaManager.resetPeers(groupId, fromAssignments(forcedAssignments.nodes()));
+                    }
+
+                    // Telling the caller that the replica is started.
+                    return trueCompletedFuture();
```

Comment:
And what if we are not in force assignments and should the replica?

Reply:
I'm not sure I understand the question correctly, but both ASSIGNED and PRIMARY_ONLY states assume that the replica is already started, so we return true here (the same as startOperation was called and returned completable future with `true` result)